src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteRuntimeResolver.cs

```diff
@@ -118,19 +115,24 @@ private object VisitCache(ServiceCallSite callSite, RuntimeResolverContext conte
 
         private object ResolveService(ServiceCallSite callSite, RuntimeResolverContext context, RuntimeResolverLock lockType, ServiceProviderEngineScope serviceProviderEngine)
         {
-            Dictionary<ServiceCacheKey, object> resolvedServices = serviceProviderEngine.ResolvedServices;
-            if (!resolvedServices.TryGetValue(callSite.Cache.Key, out object resolved))
-            {
-                resolved = VisitCallSiteMain(callSite, new RuntimeResolverContext
-                {
-                    Scope = serviceProviderEngine,
-                    AcquiredLocks = context.AcquiredLocks | lockType
-                });
+            // Note: This method has already taken lock by the caller for resolution and access synchronization.
```

Comment:
I wonder if there is a `Debug.Assert` we could add here to ensure this comment holds true. Maybe something like:

```C#
Debug.Assert(
    (lockType == RuntimeResolverLock.Root && resolvedServices is ConcurrentDictionary) ||
    (lockType == RuntimeResolverLock.Scope && Monitor.IsEntered(resolvedServices));
```